src/Microsoft.ML.Data/Transforms/CopyColumnsTransform.cs

```diff
@@ -3,25 +3,114 @@
-using System.Reflection;
+using System.Collections.Generic;
+using System.Linq;
 using System.Text;
+using Microsoft.ML.Core.Data;
 using Microsoft.ML.Runtime;
 using Microsoft.ML.Runtime.CommandLine;
 using Microsoft.ML.Runtime.Data;
 using Microsoft.ML.Runtime.EntryPoints;
 using Microsoft.ML.Runtime.Internal.Utilities;
 using Microsoft.ML.Runtime.Model;
 
-[assembly: LoadableClass(CopyColumnsTransform.Summary, typeof(CopyColumnsTransform), typeof(CopyColumnsTransform.Arguments), typeof(SignatureDataTransform),
-    CopyColumnsTransform.UserName, "CopyColumns", "CopyColumnsTransform", CopyColumnsTransform.ShortName, DocName = "transform/CopyColumnsTransform.md")]
+[assembly: LoadableClass(CopyColumnsTransformer.Summary, typeof(IDataTransform), typeof(CopyColumnsTransformer),
+    typeof(CopyColumnsTransformer.Arguments), typeof(SignatureDataTransform),
+    CopyColumnsTransformer.UserName, "CopyColumns", "CopyColumnsTransform", CopyColumnsTransformer.ShortName,
+    DocName = "transform/CopyColumnsTransformer.md")]
 
-[assembly: LoadableClass(CopyColumnsTransform.Summary, typeof(CopyColumnsTransform), null, typeof(SignatureLoadDataTransform),
-    CopyColumnsTransform.UserName, CopyColumnsTransform.LoaderSignature)]
+[assembly: LoadableClass(CopyColumnsTransformer.Summary, typeof(IDataView), typeof(CopyColumnsTransformer), null, typeof(SignatureLoadDataTransform),
+    CopyColumnsTransformer.UserName, CopyColumnsTransformer.LoaderSignature)]
+
+[assembly: LoadableClass(CopyColumnsTransformer.Summary, typeof(CopyColumnsTransformer), null, typeof(SignatureLoadModel),
+    CopyColumnsTransformer.UserName, CopyColumnsTransformer.LoaderSignature)]
+
+[assembly: LoadableClass(CopyColumnsTransformer.Summary, typeof(CopyColumnsRowMapper), null, typeof(SignatureLoadRowMapper),
+    CopyColumnsTransformer.UserName, CopyColumnsRowMapper.LoaderSignature)]
 
 namespace Microsoft.ML.Runtime.Data
 {
-    public sealed class CopyColumnsTransform : OneToOneTransformBase
+    public sealed class CopyColumnsEstimator : IEstimator<CopyColumnsTransformer>
+    {
+        private readonly (string Source, string Name)[] _columns;
+        private readonly IHost _host;
+
+        public CopyColumnsEstimator(IHostEnvironment env, string input, string output) :
+            this(env, new (string, string)[1] { (input, output) })
+        {
+        }
+
+        public CopyColumnsEstimator(IHostEnvironment env, params (string Source, string Name)[] columns)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(nameof(CopyColumnsEstimator));
+            _host.CheckValue(columns, nameof(columns));
+            var newNames = new HashSet<string>();
+            foreach (var column in columns)
+            {
+                if (newNames.Contains(column.Name))
+                    throw Contracts.ExceptUserArg(nameof(columns), $"New column {column.Name} specified multiple times");
+                newNames.Add(column.Name);
```

Comment:
Note that `HashSet<>.Add` will itself return `false` if the element was already present in the set. Therefore this check can be written more efficiently as:

```csharp
if (newNames.Add(column.Name)
    throw ...;
```